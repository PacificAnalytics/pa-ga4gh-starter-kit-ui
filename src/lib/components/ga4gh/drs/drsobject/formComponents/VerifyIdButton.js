import React, { useEffect } from 'react';
import { 
    Typography,
    Button,
    IconButton,
    Tooltip
} from '@material-ui/core';
import CheckCircle from '@material-ui/icons/CheckCircle';
import Cancel from '@material-ui/icons/Cancel';
import FormViewType from '../../../../../model/common/FormViewType';
import ApiCaller from '../../../../../utils/ApiCaller';

/*
    Used to verify if the ID entered for a parent or child DRS Object is valid.
    When the button is clicked, a GET request for the given ID will be made. If
    an error is returned, the object is set as invalid. If an object is
    successfully returned, it is valid, unless the property to be updated is
    "drs_object_parents" and the DRS Object returned is an existing blob-type
    object. Since blob-type objects cannot have children, they cannot be set as
    parent objects. If the ID has been verified and is valid, it will display a
    blue checkmark icon and the name field will be automatically populated.
    If the ID has been verified and is invalid, a red "X" icon will be displayed
    and the name field will not be populated. If the ID field is empty when the 
    "Verify" button is clicked, the ID will automatically be marked as invalid, 
    without making the GET request.
*/
const VerifyIdButton = props => {

    const defaultHandleValid = responseData => {
        props.setRelativeValid();
        props.setRelativeName(responseData.name);
    }
    const handleValid = props.customApiCallSuccessCallback ? props.customApiCallSuccessCallback : defaultHandleValid;

    const callAndValidate = relative => {
        if (relative.id === '' || relative.id === props.activeDrsObjectId) {
            props.setRelativeInvalid();
        } else {
            let requestConfig = {
                url: `${props.adminURL}/admin/ga4gh/drs/v1/objects/${relative.id}`,
                method: 'GET'
            }
            ApiCaller(requestConfig, handleValid, props.setRelativeInvalid);
        }
    }

    useEffect (() => {
        if(props.formViewType === FormViewType.EDIT && props.relative.id) {
            callAndValidate(props.relative);
        }
    }, [])

    return (
        <div>
            <Button
                aria-label='verify-id'
                color='primary'
                disabled={props.readOnly}
                variant='contained'
                size='small'
                onClick={() => callAndValidate(props.relative)}
            >
                <Tooltip title='Submit this ID for verification.'>
                    <Typography variant='button'>Verify ID</Typography>
                </Tooltip>
            </Button>

            {props.relative.isValid === true
                ?
                    <IconButton color='primary'>
                        <Tooltip title='This is a valid ID.'>
                            <CheckCircle/>
                        </Tooltip>
                    </IconButton>
                : null
            }

            {props.relative.isValid === false
                ?
                    <IconButton color='secondary'>
                        <Tooltip title='This is an invalid ID. Please enter a valid ID before proceeding.'>
                            <Cancel/>
                        </Tooltip>
                    </IconButton>
                : null
            }
        </div>
    )
}

export default VerifyIdButton;
