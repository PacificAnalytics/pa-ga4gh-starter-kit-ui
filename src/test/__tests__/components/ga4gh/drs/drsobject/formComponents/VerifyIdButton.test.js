import React from 'react';
import renderer from 'react-test-renderer';
import {render, screen} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import userEvent from '@testing-library/user-event';
import VerifyIdButton from '../../../../../../../lib/components/ga4gh/drs/drsobject/formComponents/VerifyIdButton';
import {
    mockAddObjectToList, 
    mockRemoveListItem, 
    mockUpdateScalar, 
    mockUpdateListString, 
    mockUpdateListObjectProperty, 
    mockResetListObjectProperty
} from '../../../../../../resources/MockFunctions';
import {
    mockBundleDrsObject, 
    mockBlobDrsObject
} from '../../../../../../resources/MockData';
import axios from 'axios';

jest.mock('axios');

let mockFunctions = null;

beforeEach(() => {
    mockFunctions = {
        setRelativeValid: mockResetListObjectProperty,
        setRelativeInvalid: mockResetListObjectProperty,
        setRelativeName: mockUpdateListObjectProperty
    }
    
});

afterEach(() => {
    mockFunctions.setRelativeValid.mockClear();
    mockFunctions.setRelativeInvalid.mockClear();
    mockFunctions.setRelativeValid.mockClear();
    mockResetListObjectProperty.mockClear();
    mockUpdateListObjectProperty.mockClear();
})

test('<VerifyIdButton/> should display valid relatives', () => {
    let {container} = render(
        <VerifyIdButton 
            relative={{id: mockBlobDrsObject.id, isValid: true}} 
            activeDrsObjectId='1234567890' 
            formViewType='NEW'
            {...mockFunctions}  
        />
    );
    expect(container.firstChild).toMatchSnapshot();
    expect(screen.getByTitle('This is a valid ID.')).toBeInTheDocument();
    expect(screen.queryByTitle('This is an invalid ID. Please enter a valid ID before proceeding.')).not.toBeInTheDocument();
    expect(screen.getByTitle('Submit this ID for verification.')).toBeInTheDocument();
});

test('<VerifyIdButton/> should display invalid relatives', () => {
    let {container} = render(
        <VerifyIdButton 
            relative={{id: 'abc123', isValid: false}} 
            activeDrsObjectId='1234567890' 
            formViewType='NEW'
            {...mockFunctions}  
        />
    );
    expect(container.firstChild).toMatchSnapshot();
    expect(screen.queryByTitle('This is a valid ID.')).not.toBeInTheDocument();
    expect(screen.getByTitle('This is an invalid ID. Please enter a valid ID before proceeding.')).toBeInTheDocument();
    expect(screen.getByTitle('Submit this ID for verification.')).toBeInTheDocument();
});

test('NEW <VerifyIdButton/> does not automatically validate relatives with undefined validity', () => {
    let {container} = render(
        <VerifyIdButton 
            relative={{id: 'abc123', isValid: undefined}} 
            activeDrsObjectId='1234567890' 
            formViewType='NEW'
            {...mockFunctions} 
        />
    );
    expect(container.firstChild).toMatchSnapshot();
    expect(screen.queryByTitle('This is a valid ID.')).not.toBeInTheDocument();
    expect(screen.queryByTitle('This is an invalid ID. Please enter a valid ID before proceeding.')).not.toBeInTheDocument();
    expect(screen.getByTitle('Submit this ID for verification.')).toBeInTheDocument();
});

test('EDIT <VerifyIdButton/> automatically validates relatives with undefined validity', () => {
    axios.mockResolvedValue({data: mockBlobDrsObject});
    // axios.mockImplementation(() => Promise.resolve({data: mockBlobDrsObject}));
    let {container} = render(
        <VerifyIdButton 
            relative={{id: mockBlobDrsObject.id, isValid: undefined}} 
            activeDrsObjectId='1234567890' 
            formViewType='EDIT' 
            {...mockFunctions} 
        />
    );
    expect(container.firstChild).toMatchSnapshot();
    //expect(screen.getByTitle('This is a valid ID.')).toBeInTheDocument();
    expect(screen.queryByTitle('This is an invalid ID. Please enter a valid ID before proceeding.')).not.toBeInTheDocument();
    expect(screen.getByTitle('Submit this ID for verification.')).toBeInTheDocument();
    // TODO: need to assert whether validation occurs correctly using mock function
    /* expect(mockFunctions.setRelativeValid).toHaveBeenCalled();
    expect(mockFunctions.setRelativeName).toHaveBeenCalled(); */
});

test('<VerifyIdButton/> handles relative IDs that match the active DRS Object ID', () => {
    let {container} = render(
        <VerifyIdButton  
            relative={{id: '1234567890', isValid: undefined}} 
            activeDrsObjectId='1234567890' 
            formViewType='NEW'
            {...mockFunctions} 
        />
    );
    expect(container.firstChild).toMatchSnapshot();
    let verifyIdButton = screen.getByTitle('Submit this ID for verification.');
    expect(screen.queryByTitle('This is a valid ID.')).not.toBeInTheDocument();
    expect(screen.queryByTitle('This is an invalid ID. Please enter a valid ID before proceeding.')).not.toBeInTheDocument();
    expect(verifyIdButton).toBeInTheDocument();
    userEvent.click(verifyIdButton);
    expect(mockFunctions.setRelativeInvalid.mock.calls.length).toBe(1);
});

test('<VerifyIdButton/> handles empty relative IDs', () => {
    let {container} = render(
        <VerifyIdButton 
            relative={{id: '', isValid: undefined}} 
            activeDrsObjectId='1234567890' 
            formViewType='NEW'
            {...mockFunctions} 
        />
    );
    expect(container.firstChild).toMatchSnapshot();
    let verifyIdButton = screen.getByTitle('Submit this ID for verification.');
    expect(screen.queryByTitle('This is a valid ID.')).not.toBeInTheDocument();
    expect(screen.queryByTitle('This is an invalid ID. Please enter a valid ID before proceeding.')).not.toBeInTheDocument();
    expect(verifyIdButton).toBeInTheDocument();
    userEvent.click(verifyIdButton);
    expect(mockFunctions.setRelativeInvalid.mock.calls.length).toBe(1);
})