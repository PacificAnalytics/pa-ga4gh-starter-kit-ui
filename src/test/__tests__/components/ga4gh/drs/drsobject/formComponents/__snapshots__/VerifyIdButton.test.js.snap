// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VerifyIdButton/> handles empty relative IDs 1`] = `
<div>
  <button
    aria-label="verify-id"
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      <span
        class="MuiTypography-root MuiTypography-button"
        title="Submit this ID for verification."
      >
        Verify ID
      </span>
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</div>
`;

exports[`<VerifyIdButton/> handles relative IDs that match the active DRS Object ID 1`] = `
<div>
  <button
    aria-label="verify-id"
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      <span
        class="MuiTypography-root MuiTypography-button"
        title="Submit this ID for verification."
      >
        Verify ID
      </span>
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</div>
`;

exports[`<VerifyIdButton/> should display invalid relatives 1`] = `
<div>
  <button
    aria-label="verify-id"
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      <span
        class="MuiTypography-root MuiTypography-button"
        title="Submit this ID for verification."
      >
        Verify ID
      </span>
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
  <button
    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSecondary"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiIconButton-label"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root"
        focusable="false"
        title="This is an invalid ID. Please enter a valid ID before proceeding."
        viewBox="0 0 24 24"
      >
        <path
          d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
        />
      </svg>
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</div>
`;

exports[`<VerifyIdButton/> should display valid relatives 1`] = `
<div>
  <button
    aria-label="verify-id"
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      <span
        class="MuiTypography-root MuiTypography-button"
        title="Submit this ID for verification."
      >
        Verify ID
      </span>
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
  <button
    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiIconButton-label"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root"
        focusable="false"
        title="This is a valid ID."
        viewBox="0 0 24 24"
      >
        <path
          d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
        />
      </svg>
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</div>
`;

exports[`EDIT <VerifyIdButton/> automatically validates relatives with undefined validity 1`] = `
<div>
  <button
    aria-label="verify-id"
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      <span
        class="MuiTypography-root MuiTypography-button"
        title="Submit this ID for verification."
      >
        Verify ID
      </span>
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</div>
`;

exports[`NEW <VerifyIdButton/> does not automatically validate relatives with undefined validity 1`] = `
<div>
  <button
    aria-label="verify-id"
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      <span
        class="MuiTypography-root MuiTypography-button"
        title="Submit this ID for verification."
      >
        Verify ID
      </span>
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</div>
`;
